File: test/ut/cpp/testUtils.cpp
/* -------------------------------------------------------------------------
 * This file is part of the MindStudio project.
 * Copyright (c) 2025 Huawei Technologies Co.,Ltd.
 *
 * MindStudio is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *
 *          http://license.coscl.org.cn/MulanPSL2
 *
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * -------------------------------------------------------------------------
*/

#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include <string>
#include <string>
#include <set>

#include "msServiceProfiler/Utils.h"

using namespace ::testing;
using namespace MsUtils;

TEST(ServiceProfilerUtilsTest, SplitStrSuccess)
{
    std::string input = "Hello World";
    char splitChar = ' ';
    auto result = MsUtils::SplitStr(input, splitChar);
    EXPECT_EQ(result.first, "Hello");
    EXPECT_EQ(result.second, "World");
}

TEST(ServiceProfilerUtilsTest, SplitStrWithNoSplitChar) {
    std::string input = "HelloWorld";
    char splitChar = ' ';
    auto result = MsUtils::SplitStr(input, splitChar);
    EXPECT_EQ(result.first, "HelloWorld");
    EXPECT_EQ(result.second, "");
}

TEST(ServiceProfilerUtilsTest, SplitStrToSetSuccess)
{
    std::set<std::string> filterSet;
    const char *name = "example1;example2;example3";
    char splitSymbol = ';';
    auto result = SplitStringToSet(name, splitSymbol);
    EXPECT_EQ(result.size(), 3);
    EXPECT_NE(result.find("example1"), result.end());
    EXPECT_NE(result.find("example2"), result.end());
    EXPECT_NE(result.find("example3"), result.end());
}

TEST(ServiceProfilerUtilsTest, SplitStrToSetSuccessWithEmptySubstrings)
{
    std::set<std::string> filterSet;
    const char *name = "example1;example2;;example3;";
    char splitSymbol = ';';
    auto result = SplitStringToSet(name, splitSymbol);
    EXPECT_EQ(result.size(), 3);
    EXPECT_NE(result.find("example1"), result.end());
    EXPECT_NE(result.find("example2"), result.end());
    EXPECT_NE(result.find("example3"), result.end());
}

TEST(ServiceProfilerUtilsTest, SplitStrToSetSuccessWithOneSubstring)
{
    std::set<std::string> filterSet;
    const char *name = "example";
    char splitSymbol = ';';
    auto result = SplitStringToSet(name, splitSymbol);
    EXPECT_EQ(result.size(), 1);
    EXPECT_NE(result.find("example"), result.end());
}

File: test/ut/python/test_ms_service_metric/test_utils.py
# -------------------------------------------------------------------------
# This file is part of the MindStudio project.
# Copyright (c) 2025 Huawei Technologies Co.,Ltd.
#
# MindStudio is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#
#          http://license.coscl.org.cn/MulanPSL2
#
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# -------------------------------------------------------------------------

import pytest

from ms_service_metric.utils.function_context import FunctionContext


class TestFunctionContextGivenValidData:
    def test_when_create_context_then_local_values_is_none(self):
        ctx = FunctionContext()
        assert ctx.local_values is None
        assert ctx.return_value is None

    def test_when_set_local_values_then_values_are_stored(self):
        ctx = FunctionContext()
        ctx.local_values = {"x": 1, "y": 2}
        assert ctx.local_values == {"x": 1, "y": 2}

    def test_when_set_return_value_then_value_is_stored(self):
        ctx = FunctionContext()
        ctx.return_value = "test_result"
        assert ctx.return_value == "test_result"

    def test_when_get_existing_key_then_return_value(self):
        ctx = FunctionContext()
        ctx.local_values = {"key": "value"}
        assert ctx.get("key") == "value"

    def test_when_get_nonexistent_key_with_default_then_return_default(self):
        ctx = FunctionContext()
        ctx.local_values = {}
        assert ctx.get("missing", "default") == "default"

    def test_when_get_nonexistent_key_without_default_then_return_none(self):
        ctx = FunctionContext()
        ctx.local_values = {}
        assert ctx.get("missing") is None

    def test_when_getitem_with_existing_key_then_return_value(self):
        ctx = FunctionContext()
        ctx.local_values = {"key": "value"}
        assert ctx["key"] == "value"

    def test_when_getitem_with_nonexistent_key_then_raise_key_error(self):
        ctx = FunctionContext()
        ctx.local_values = {}
        with pytest.raises(KeyError):
            _ = ctx["missing"]

    def test_when_contains_with_existing_key_then_return_true(self):
        ctx = FunctionContext()
        ctx.local_values = {"key": "value"}
        assert "key" in ctx

    def test_when_contains_with_nonexistent_key_then_return_false(self):
        ctx = FunctionContext()
        ctx.local_values = {}
        assert "missing" not in ctx

    def test_when_contains_with_none_local_values_then_return_false(self):
        ctx = FunctionContext()
        assert "key" not in ctx

    def test_when_get_with_none_local_values_then_return_default(self):
        ctx = FunctionContext()
        assert ctx.get("key", "default") == "default"

